Doc/c-api/typeobj.rst

```diff
@@ -654,16 +654,31 @@ and :c:type:`PyType_Type` effectively act as defaults.)
    the instance is still in existence, but there are no references to it.  The
    destructor function should free all references which the instance owns, free all
    memory buffers owned by the instance (using the freeing function corresponding
-   to the allocation function used to allocate the buffer), and finally (as its
-   last action) call the type's :c:member:`~PyTypeObject.tp_free` function.  If the type is not
-   subtypable (doesn't have the :const:`Py_TPFLAGS_BASETYPE` flag bit set), it is
+   to the allocation function used to allocate the buffer), and call the type's
+   :c:member:`~PyTypeObject.tp_free` function.  If the type is not subtypable
+   (doesn't have the :const:`Py_TPFLAGS_BASETYPE` flag bit set), it is
    permissible to call the object deallocator directly instead of via
    :c:member:`~PyTypeObject.tp_free`.  The object deallocator should be the one used to allocate the
    instance; this is normally :c:func:`PyObject_Del` if the instance was allocated
    using :c:func:`PyObject_New` or :c:func:`PyObject_VarNew`, or
    :c:func:`PyObject_GC_Del` if the instance was allocated using
    :c:func:`PyObject_GC_New` or :c:func:`PyObject_GC_NewVar`.
 
+   Finally, if the type is heap allocated (:const:`Py_TPFLAGS_HEAPTYPE`), the
+   deallocator should decrement the reference count for its type object after
+   calling the type deallocator. In order to avoid dangling pointers, the
+   recommended way to achieve this is:
+
+   .. code-block:: c
+
+     static void foo_dealloc(foo_object *self) {
+         PyTypeObject *tp = Py_TYPE(self);
+         // free references and buffers here
+         tp->tp_free(self);
```

Comment:
Since this is a heap type, change this to use the slot API to be compliant with PEP384.

```
freefunc free_func = PyType_GetSlot(tp, Py_tp_free);
free_func(self);
```

Reply:
[PEP 384](https://www.python.org/dev/peps/pep-0384/) defines a stable ABI – an API subset you can use when you can't recompile extensions for each minor version of CPython.

How is it relevant here?

Reply:
@encukou it is very relevant! What you said is correct, but this PEP has the added benefit of making `PyTypeObject` opaque. This gives the type system control over the implementation of the type object. For instance, in our system, the type object has a very different implementation, yet, we can correctly interface with C Extensions. Using `tp_free` directly ties the code to CPython's implementation.

By the way, this is only relevant for C extensions, that is, `cpython/Modules` + third-party libs. Internal code CPython code should be free to use all the implementation details it wants! Given that the updated doc has the context of heap types, it's a fair assumption to make :-)

Reply:
I agree that using the stable ABI is good, and we all migrate to it eventually. However, it's orthogonal to this change, which is about making the documentation *correct*. Pushing a stable ABI agenda in a fix for "Clarify that tp_dealloc must decref for heap allocated type" seems too sneaky for me – even though I count myself as a proponent of the stable ABI :)

In CPython, `PyType_GetSlot` is likely* much slower than using the slot directly, so it is not a clear win.
I'm for avoiding direct slot access (both in stdlib and extension modules), but only after a wider discussion and perhaps after also adding a fast (and unstable) API.

And again: the text above talks about `PyTypeObject.tp_free`. The example is confusingly out of sync with that. I think we should change both occurences, but please, do it in a different PR.

---

* (I haven't measured, but there's a bunch of extra checks in CPython's `PyType_GetSlot`)

Reply:
Reverted back to `tp->tp_free(self);`

Reply:
Woops, didn't mean to try any sneaky stuff! 😛 

I agree with your point, let's get to anything related to PEP384 independently.